<a href="https://colab.research.google.com/github/https-deeplearning-ai/tensorflow-1-public/blob/main/C1/W2/ungraded_labs/C1_W2_Lab_2_callbacks.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ungraded Lab: Using Callbacks to Control Training

In this lab, you will use the [Callbacks API](https://keras.io/api/callbacks/) to stop training when a specified metric is met. This is a useful feature so you won't need to complete all epochs when this threshold is reached. For example, if you set 1000 epochs and your desired accuracy is already reached at epoch 200, then the training will automatically stop. Let's see how this is implemented in the next sections.


## Load and Normalize the Fashion MNIST dataset

Like the previous lab, you will use the Fashion MNIST dataset again for this exercise. And also as mentioned before, you will normalize the pixel values to help optimize the training.

In [1]:
import tensorflow as tf

# If you want to run this code VSCode or PyCharm, use this code segment.
# TensorFlow GPU won't support for versions of 12.0 and higher.
import os
os.environ['KMP_DUPLICATE_LIB_OK']='True'

# Instantiate the dataset API
fmnist = tf.keras.datasets.fashion_mnist

# Load the dataset
(x_train, y_train),(x_test, y_test) = fmnist.load_data()
"""
x_train: train verileri
y_train : train etiketleri
x_test : test verileri
y_test : test etiketleri
"""

"""
? Şimdi, bunlar nelerdir?
- Böyle bir sinir ağını kurarken, modelin görüntüleri tanımada ne kadar iyi olduğunu test etmek için henüz
görmediği sinit ağını ve benzer verileri eğitmek için verilerinizin bir kısmını kullanmak güzel bir stratejidir.

Dolayısıyla Fashion-MNIST veri setinde , 70.000 görüntünün 60.000'ı ağı eğitmek için kullanılır ve daha sonra
daha önceden görmediği 10.000 gröüntü ne kadar iyi, ne kadar kötü performans gösterdiğini test etmek için
kullanılabilir.
"""

# Normalize the pixel values
x_train, x_test = x_train / 255.0, x_test / 255.0

### The Structure of Fashion MNIST data

- The image is represented as a 28x28 array of greyscales, and how its label is a number.
- Using a number is a first step in avoiding bias -- instead of labelling it with words in a specific language and excluding people who don’t speak that language!

## Creating a Callback class

You can create a callback by defining a class that inherits the [tf.keras.callbacks.Callback](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback) base class. From there, you can define available methods to set where the callback will be executed. For instance below, you will use the [on_epoch_end()](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback#on_epoch_end) method to check the loss at each training epoch.

In [2]:
class myCallback(tf.keras.callbacks.Callback):
  def on_epoch_end(self, epoch, logs={}):
    '''
    Halts the training after reaching 60 percent accuracy

    Args:
      epoch (integer) - index of epoch (required but unused in the function definition below)
      logs (dict) - metric results from the training epoch
    '''

    # Check accuracy
    if(logs.get('loss') < 0.4):

      # Stop if threshold is met
      print("\nLoss is lower than 0.4 so cancelling training!")
      self.model.stop_training = True

# Instantiate class
callbacks = myCallback()

## Define and compile the model

Next, you will define and compile the model. The architecture will be similar to the one you built in the previous lab. Afterwards, you will set the optimizer, loss, and metrics that you will use for training.

In [3]:
# Define the model
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(512, activation=tf.nn.relu),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])

# Compile the model
model.compile(optimizer=tf.optimizers.Adam(),
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])



### Train the model

Now you are ready to train the model. To set the callback, simply set the `callbacks` parameter to the `myCallback` instance you declared before. Run the cell below and observe what happens.

In [4]:
# Train the model with a callback
model.fit(x_train, y_train, epochs=10, callbacks=[callbacks])

Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.4718 - accuracy: 0.8310
Epoch 2/10
1874/1875 [============================>.] - ETA: 0s - loss: 0.3598 - accuracy: 0.8691
Loss is lower than 0.4 so cancelling training!
1875/1875 [==============================] - 6s 3ms/step - loss: 0.3597 - accuracy: 0.8691


You will notice that the training does not need to complete all 10 epochs. By having a callback at each end of the epoch, it is able to check the training parameters and compare if it meets the threshold you set in the function definition. In this case, it will simply stop when the loss falls below `0.40` after the current epoch.

*Optional Challenge: Modify the code to make the training stop when the accuracy metric exceeds 60%.*

That concludes this simple exercise on callbacks!